File: src/ros2_bridge/mappers/geometry_msgs/transform_stamped.rs
//! Typed mapper for `geometry_msgs/msg/TransformStamped`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn transform_stamped_to_bus(msg: ros_env::geometry_msgs::msg::TransformStamped) -> crate::geometry_msgs::msg::v1::TransformStamped {
    crate::geometry_msgs::msg::v1::TransformStamped {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        child_frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.child_frame_id),
        transform: Some(crate::ros2_bridge::mappers::geometry_msgs::transform::transform_to_bus(msg.transform)),
    }
}

pub(crate) fn transform_stamped_to_ros(bus: crate::geometry_msgs::msg::v1::TransformStamped) -> ros_env::geometry_msgs::msg::TransformStamped {
    ros_env::geometry_msgs::msg::TransformStamped {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        child_frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.child_frame_id),
        transform: crate::ros2_bridge::mappers::geometry_msgs::transform::transform_to_ros(bus.transform.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsTransformStampedMapper;

impl TypedTopicMapper for GeometryMsgsTransformStampedMapper {
    type Ros = ros_env::geometry_msgs::msg::TransformStamped;
    type Bus = crate::geometry_msgs::msg::v1::TransformStamped;

    fn type_name(&self) -> &'static str {
        "geometry_msgs/msg/TransformStamped"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(transform_stamped_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(transform_stamped_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/transform_stamped.py
"""Generated mapper for `geometry_msgs/msg/TransformStamped`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.transform import transform_to_bus, transform_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.geometry_msgs.msg.v1 import TransformStamped as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def transform_stamped_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.child_frame_id = str(msg.child_frame_id)
    bus.transform.CopyFrom(transform_to_bus(msg.transform))
    return bus


def transform_stamped_to_ros(bus):
    from geometry_msgs.msg import TransformStamped as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.child_frame_id = str(bus.child_frame_id)
    out.transform = transform_to_ros(bus.transform)
    return out


class GeometryMsgsTransformStampedMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from geometry_msgs.msg import TransformStamped as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return transform_stamped_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return transform_stamped_to_ros(bus)
